### PR TITLE
[SYCL] Fix ICE when -g and -ffile-prefix-map flags are set

### PR DESCRIPTION
**Pull request by zahiraam**
Set the source file name when footer is present and option fmacro-prefix-map is used when generating debug information.

This patch is to fix a crash in the compiler when generating debug information with the option -ffile-prefix-map.
Example:
$ pwd
/tmp/tmp.jocQk9WVa0
$
$ cat builtin.cpp 
void a(__builtin_va_list);
using ::a;
$
$ clang -c -g -fsycl /tmp/tmp.jocQk9WVa0/builtin.cpp
$ 
$ clang -g -fsycl -ffile-prefix-map=/tmp/= /tmp/tmp.jocQk9WVa0/builtin.cpp 
Expected<T> must be checked before access or destruction.
Unchecked Expected<T> contained error:
No such file or directoryPLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: 
…………………/tmp/tmp.jocQk9WVa0/builtin.cpp
1.	<eof> parser at end of file
2.	/tmp/tmp.jocQk9WVa0/builtin.cpp:2:9: LLVM IR generation of declaration 'a

https://github.com/intel/llvm/blob/sycl/clang/lib/CodeGen/CGDebugInfo.cpp#L397 is returning a error because the FileName is wrong.

### CODE REVIEW

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 1]
```diff
@@ -411,6 +411,14 @@ llvm::DIFile *CGDebugInfo::getOrCreateFile(SourceLocation Loc) {
     // createFile() below for canonicalization if the source file was specified
     // with an absolute path.
     FileName = TheCU->getFile()->getFilename();
+    if (CGM.getCodeGenOpts().SYCLUseMainFileName &&
+        CGM.getLangOpts().MacroPrefixMap.size() > 0) {
+      // When fmacro-prefix-map is used, the original source file
+      // file name is indicated by FullMainFileName instead of the CU.
+      auto &CGO = CGM.getCodeGenOpts();
+      FileName = CGO.FullMainFileName;
```
Comment (elizabethandrews) [inline clang/lib/CodeGen/CGDebugInfo.cpp:419]: What is FullMainFileName?

File: clang/test/CodeGenSYCL/debug-info-file-prefix-map.cpp [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// RUN: %clang_cc1 -triple spir64-unknown-unknown -fsycl-is-device \
+// RUN: -fsycl-int-header=%t.header.h -fsycl-int-footer=%t.footer.h \
+// RUN: -main-file-name %S/Inputs/builtin.cpp -fsycl-use-main-file-name \
+// RUN: -full-main-file-name "%S/Inputs/builtin.cpp" \
+// RUN: -fmacro-prefix-map=%S/Inputs/= -fcoverage-prefix-map=%S/Inputs/= \
+// RUN: -fdebug-prefix-map=%S/Inputs/= \
+// RUN: -debug-info-kind=constructor -emit-llvm -O0 -o - "%S/Inputs/builtin.cpp" \
+// RUN: | FileCheck %s
```
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/debug-info-file-prefix-map.cpp:8]: Please add a description for this test

**Review by elizabethandrews — COMMENTED** [review]
Can you describe what this change is doing in more detail in PR description?

File: clang/test/CodeGenSYCL/debug-info-file-prefix-map.cpp (hunk 2 above)
Comment (zahiraam) [reply clang/test/CodeGenSYCL/debug-info-file-prefix-map.cpp:8]: FullMainFileName is declared here [llvm/ llvm/CGDebugInfo.cpp at sycl · intel/llvm · GitHubCGDebugInfo.cpp at sycl · intel/llvm · GitHub](https://github.com/intel/llvm/blob/sycl/clang/lib/CodeGen/CGDebugInfo.cpp#L551) and represents the name of the original source file name.

Comment (zahiraam) [reply clang/test/CodeGenSYCL/debug-info-file-prefix-map.cpp:8]: It's a codegen option triggered by https://github.com/intel/llvm/blob/sycl/clang/include/clang/Driver/Options.td#L6333

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 3]
```diff
@@ -411,6 +411,14 @@ llvm::DIFile *CGDebugInfo::getOrCreateFile(SourceLocation Loc) {
     // createFile() below for canonicalization if the source file was specified
     // with an absolute path.
     FileName = TheCU->getFile()->getFilename();
```
Comment (premanandrao) [inline clang/lib/CodeGen/CGDebugInfo.cpp:413]: Can you move the above assignment to an else clause?

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 4]
```diff
@@ -411,6 +411,14 @@ llvm::DIFile *CGDebugInfo::getOrCreateFile(SourceLocation Loc) {
...
+        CGM.getLangOpts().MacroPrefixMap.size() > 0) {
+      // When fmacro-prefix-map is used, the original source file
+      // file name is indicated by FullMainFileName instead of the CU.
+      auto &CGO = CGM.getCodeGenOpts();
+      FileName = CGO.FullMainFileName;
+      FID = ComputeValidFileID(SM, CGO.FullMainFileName);
+    }
   } else {
     PresumedLoc PLoc = SM.getPresumedLoc(Loc);
     FileName = PLoc.getFilename();
```
Comment (premanandrao) [inline clang/lib/CodeGen/CGDebugInfo.cpp:424]: Line 429 below is not needed.

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 5]
```diff
@@ -407,18 +407,25 @@ llvm::DIFile *CGDebugInfo::getOrCreateFile(SourceLocation Loc) {
...
+      // specified with an absolute path.
+      FileName = TheCU->getFile()->getFilename();
+    }
   } else {
     PresumedLoc PLoc = SM.getPresumedLoc(Loc);
     FileName = PLoc.getFilename();
 
     if (FileName.empty()) {
       FileName = TheCU->getFile()->getFilename();
-    } else {
```
Comment (elizabethandrews) [inline clang/lib/CodeGen/CGDebugInfo.cpp:420]: Why was this removed?

File: clang/test/CodeGenSYCL/debug-info-file-prefix-map.cpp (hunk 2 above)
Comment (elizabethandrews) [reply clang/test/CodeGenSYCL/debug-info-file-prefix-map.cpp:8]: Sorry if I wasn't clear earlier. I meant a description in this file. All test files need a comment describing what the test does.

File: clang/lib/CodeGen/CGDebugInfo.cpp (hunk 5 above)
Comment (zahiraam) [reply clang/lib/CodeGen/CGDebugInfo.cpp:420]: This was a simplification that Prem had proposed in the comments above.

**Comment by zahiraam** [comment]
@bwyma Thanks for reviewing and checking the debug information. I have fixed the issue you raised about the directory path. Also added in the other LIT tests a check line for the directory to make sure it's correct (instead of using regular expression). Would you please check if this new patch fixes the issue? Thanks.

File: clang/lib/CodeGen/CGDebugInfo.cpp (hunk 5 above)
Comment (elizabethandrews) [reply clang/lib/CodeGen/CGDebugInfo.cpp:420]: I'm not sure if I am missing something. What is FileName set to if `!FileName.empty()`

File: clang/lib/CodeGen/CGDebugInfo.cpp [hunk 6]
```diff
@@ -656,9 +663,13 @@ void CGDebugInfo::CreateCompileUnit() {
   // file was specified with an absolute path.
   if (CSKind)
     CSInfo.emplace(*CSKind, Checksum);
-  llvm::DIFile *CUFile = DBuilder.createFile(
-      remapDIPath(MainFileName), remapDIPath(getCurrentDirname()), CSInfo,
-      getSource(SM, SM.getMainFileID()));
+
+  if (!CGM.getCodeGenOpts().SYCLUseMainFileName)
```
Comment (elizabethandrews) [inline clang/lib/CodeGen/CGDebugInfo.cpp:667]: Please add comments describing your change. Is the comment on L660 still accurate?

**Comment by zahiraam** [comment]
> LGTM.

Thanks.

File: clang/lib/CodeGen/CGDebugInfo.cpp (hunk 5 above)
Comment (zahiraam) [reply clang/lib/CodeGen/CGDebugInfo.cpp:420]: FileName is defined at line 425 above. If it's empty we computed it from the TheCU, otherwise we keep going.

File: clang/test/CodeGenSYCL/Inputs/builtin.cpp [hunk 7]
```diff
@@ -0,0 +1,3 @@
+void a(__builtin_va_list);
+using ::a;
+
```
Comment (bader) [inline clang/test/CodeGenSYCL/Inputs/builtin.cpp:3]: I believe we can remove this file and move the content to `clang/test/CodeGenSYCL/debug-info-file-prefix-map.cpp` directly.

**Review by smanna12 — APPROVED** [review]

File: clang/lib/CodeGen/CGDebugInfo.cpp (hunk 5 above)
Comment (premanandrao) [reply clang/lib/CodeGen/CGDebugInfo.cpp:420]: @elizabethandrews, the else clause was setting it to the same value that the assignment on line 425 was doing, so it was redundant.

**Review by premanandrao — APPROVED** [review]

**Comment by zahiraam** [comment]
@intel/dpcpp-devops-reviewers I have been re-basing and re-testing this PR for a week. Can someone help with the latest fail please? Thanks.
It says:
"This request was automatically failed because there were no enabled runners online to process the request for more than 1 days."

**Comment by aelovikov-intel** [comment]
@apstasen , could you take a look please? We've seen similar issue with not getting AWS CUDA runner picked up in at least one another PR.

**Comment by bader** [comment]
> @intel/dpcpp-devops-reviewers I have been re-basing and re-testing this PR for a week. Can someone help with the latest fail please? Thanks. It says: "This request was automatically failed because there were no enabled runners online to process the request for more than 1 days."

I hope https://github.com/intel/llvm/pull/8567 fixes this issue.

**Comment by zahiraam** [comment]
> > @intel/dpcpp-devops-reviewers I have been re-basing and re-testing this PR for a week. Can someone help with the latest fail please? Thanks. It says: "This request was automatically failed because there were no enabled runners online to process the request for more than 1 days."
> 
> I hope #8567 fixes this issue.

The testing started  14h 14m 8s ago. Unlless I am mistaken this is still hanging?
"Requested labels: aws-cuda_4318684912-4
Job defined at: intel/llvm/.github/workflows/sycl_linux_build_and_test.yml@refs/heads/sycl
Reusable workflow chain:
intel/llvm/.github/workflows/sycl_precommit.yml@refs/heads/sycl (fe12caba47874ac522953bfec56f740e1a823996)
-> intel/llvm/.github/workflows/sycl_linux_build_and_test.yml@refs/heads/sycl (fe12caba47874ac522953bfec56f740e1a823996)
Waiting for a runner to pick up this job..."
I seem to have no way of re-starting the test again. 
@bader, @Stasenko, Alexander P?

**Comment by bader** [comment]
@zahiraam, you need to merge the fix to your `DBI` branch.

**Comment by zahiraam** [comment]
@bader Your fix seems to have worked. Thanks.
Can we now merge in this PR?
